django-stubs/db/models/base.pyi

```diff
@@ -19,14 +19,14 @@ class ModelState:
     fields_cache: ModelStateFieldsCacheDescriptor
 
 class ModelBase(type):
-    @property
-    def objects(cls: type[_Self]) -> BaseManager[_Self]: ...  # type: ignore[misc]
     @property
     def _default_manager(cls: type[_Self]) -> BaseManager[_Self]: ...  # type: ignore[misc]
     @property
     def _base_manager(cls: type[_Self]) -> BaseManager[_Self]: ...  # type: ignore[misc]
 
 class Model(metaclass=ModelBase):
+    objects: BaseManager[Self]
```

Comment:
Changing this to `@classproperty` seems to fix the `create_animal_generic` part of issue #1653:

```suggestion
    @classproperty
    def objects(cls: type[_Self]) -> BaseManager[_Self]: ...
```

Reply:
DOH! That causes new errors to appear...

```
django-stubs/django-stubs/contrib/contenttypes/models.pyi:18: error: Incompatible types in assignment (expression has type "ContentTypeManager", base class "Model" defined the type as "Callable[[Type[_Self]], BaseManager[_Self]]")
```

Reply:
`classproperty` can’t help us with typing this: its constructor annotates the `cls` parameter as `Any`, so there’s nothing to connect `_Self` to the actual model type.

https://github.com/typeddjango/django-stubs/blob/325006ccd72a9b838cf841b1c61ba4092b084f99/django-stubs/utils/functional.pyi#L99-L101

Reply:
It does help. The `Any` doesn't matter here.

Before `classproperty` even comes into play, mypy has already resolved the relationship between the parameter and return value. The method signature `def objects(cls: type[_Self]) -> BaseManager[_Self]: ...` is evaluated first, before decorators.

But as I point out above, this breaks overriding `objects` attribute in a custom model.

Reply:
Ignoring the decorator is clearly a mypy bug. And it still doesn’t help us, because ignoring the decorator makes mypy incorrectly assume the first parameter is an instance of `Model`, not any kind of `type`.

```python
from django.utils.functional import classproperty

class Model:
    @classproperty
    def foo(cls: "Model") -> "Model":  # incorrectly allowed
        return cls

    @classproperty
    def bar(cls: "type[Model]") -> "type[Model]":  # incorrect error: The erased type of self "Type[a.Model]" is not a supertype of its class "a.Model"
        return cls

print(Model.foo is Model)  # True at runtime
reveal_type(Model)  # Revealed type is "def () -> __main__.Model"
reveal_type(Model.foo)  # Revealed type is "__main__.Model"
```